standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java

```diff
@@ -79,17 +92,21 @@
   private AtomicBoolean isDatabaseCacheDirty = new AtomicBoolean(false);
 
   // For caching TableWrapper objects. Key is aggregate of database name and table name
-  private Map<String, TableWrapper> tableCache = new TreeMap<>();
+  private ReentrantReadWriteLock tableCacheRWLock = new ReentrantReadWriteLock(true);
```

Comment:
Is this a global lock? Why we need another one in addition to cacheLock?

Reply:
Yes, it's a global lock. It serves a different purpose than the cacheLock. There can be concurrent threads trying to read from the cache, they should be able to enter the cacheLock.readLock() when current thread is modifying the tableCache. 
When modification of tableWrapper happens, we usually do the following
try {
      cacheLock.readLock().lock();
      tableCacheRWLock.writeLock().lock();
      TableWrapper tblWrapper = tableCache.getIfPresent(CacheUtils.buildTableKey(catName, dbName, tblName));
      if (tblWrapper != null) {
        tblWrapper.cachePartitions(parts, this, false);
      }
    } finally {
      tableCacheRWLock.writeLock().unlock();
      cacheLock.readLock().unlock();

My idea is the first lock, global cacheLock is there to guard read/write at the database level, and the tableCacheRWLock is there to guard the update block, so that the update to the tableWrapper.cachePartitions can happen atomically. Without this tableCacheRWLock, it'd be possible for other concurrent thread to change the tableWrapper between the time tableCache.getIfPresent( ) is called, and tblWrapper.cachePartitions(parts, this, false); is called, causing inconsistency. So using two locks, we can be sure threads are entering the cache with some lock at least (the global cacheLock), and if modification of tableWrapper is needed, need to enter write lock on tableCacheRWLock.